desc/particles.py

```diff
@@ -0,0 +1,1239 @@
+        xp, yp, zeta, vpar = x
+        # we use cartesian-like coordinates xp and yp which are x=rho*cos(theta) and
+        # y=rho*sin(theta) for integration, but convert them to flux coordinates for
+        # compute functions. This way, we don't have the problem of terminating the
+        # integration when rho<0, but actually the particle is still in the plasma
+        # volume.
+        rho = jnp.sqrt(xp**2 + yp**2)
+        theta = jnp.arctan2(yp, xp)
+        # compute functions are not correct for very small rho
+        rho = jnp.where(rho < 1e-6, 1e-6, rho)
+        iota = kwargs.get("iota", None)
+        grid = Grid(
+            jnp.array([rho, theta, zeta]).T,
+            spacing=jnp.zeros((3,)).T,
+            jitable=True,
+        )
+        data_keys = [
+            "B",
+            "|B|",
+            "grad(|B|)",
+            "e^rho",
+            "e^theta",
+            "e^zeta",
+            "b",
+        ]
+
+        transforms = get_transforms(data_keys, eq, grid, jitable=True)
+        profiles = get_profiles(data_keys, eq, grid)
+        if iota is not None:
+            profiles["iota"] = iota
+        data = compute_fun(eq, data_keys, params, transforms, profiles)
+
+        # derivative of the guiding center position in R, phi, Z coordinates
+        Rdot = vpar * data["b"] + (
+            (m / q / data["|B|"] ** 2)
+            * ((mu * data["|B|"] / m) + vpar**2)
+            * cross(data["b"], data["grad(|B|)"])
+        )
+        # take dot product for rho, theta and zeta coordinates
+        rhodot = dot(Rdot, data["e^rho"])
+        thetadot = dot(Rdot, data["e^theta"])
+        zetadot = dot(Rdot, data["e^zeta"])
+        # get the derivative for cartesian-like coordinates
+        xpdot = rhodot * jnp.cos(theta) - rho * thetadot * jnp.sin(theta)
+        ypdot = rhodot * jnp.sin(theta) + rho * thetadot * jnp.cos(theta)
+        # derivative the parallel velocity
+        vpardot = -mu / m * dot(data["b"], data["grad(|B|)"])
+        dxdt = jnp.array([xpdot, ypdot, zetadot, vpardot]).reshape(x.shape)
+        return dxdt.squeeze()
+
+    def _compute_lab_coordinates(self, x, field, params, m, q, mu, **kwargs):
+        """Compute the RHS of the ODE using MagneticField.
+
+        This function is written for vmap, so it expects x to be a coordinate of a
+        single particle, and args to be a tuple of (m, q, mu, field, params, kwargs)
+        with m, q and mu (mv⊥²/2|B|) being scalars.
+
+        kwargs can contain source_grid for the magnetic field computation.
+        """
+        vpar = x[-1]
+        coord = x[:-1]
+
+        field_norm_compute = lambda y: jnp.linalg.norm(
+            field.compute_magnetic_field(y, params=params, **kwargs), axis=-1
+        ).squeeze()
+
+        # magnetic field vector in R, phi, Z coordinates
+        B = field.compute_magnetic_field(coord, params=params, **kwargs).squeeze()
+        grad_B = Derivative(field_norm_compute, mode="grad")(coord)
```

Comment:
TODO for later: implement a way to do this for all fields, either with AD like this or through analytic derivatives on a case-by-case basis